cookbooks/aws-parallelcluster-config/templates/default/directory_service/generate_ssh_key.sh.erb

```diff
@@ -1,14 +1,23 @@
 #!/bin/bash
 set -ex
 env
-if [ ! -d /home/${PAM_USER}/.ssh ] ; then
-    mkdir /home/${PAM_USER}/.ssh
-    ssh-keygen -q -t rsa -f /home/${PAM_USER}/.ssh/id_rsa -N ''
-    cat /home/${PAM_USER}/.ssh/id_rsa.pub >> /home/${PAM_USER}/.ssh/authorized_keys
-    chmod 0600 /home/${PAM_USER}/.ssh/authorized_keys
-    ssh-keyscan <%=  node['hostname'] %> > /home/${PAM_USER}/.ssh/known_hosts
-    chmod 0600 /home/${PAM_USER}/.ssh/known_hosts
-    chown ${PAM_USER}:$(id -g ${PAM_USER}) -R /home/${PAM_USER}/.ssh
-else
-    exit 0
-fi
+
+# Root does not need SSH key generation
+[ ${PAM_USER} == "root" ] && exit 0
+
+# We should not assume that every user has its home directory in /home/$USER.
+# So, we retrieve the home directory in the most generic way.
+user_ssh_dir="$(eval echo ~${PAM_USER})/.ssh"
+
+# Skip SSH key creation if the SSH has been already configured for the user.
+# We assume that SSH has been already configured if the directory .ssh already exists in the user home.
+[ -d "${user_ssh_dir}" ] && exit 0
+
+mkdir "${user_ssh_dir}"
```

Comment:
Can be useful to enforce permission in case of different umask set.

`mkdir -m 0700 "${user_ssh_dir}"` 

Here it is sure that the home dir already exist?

Reply:
Thank for the suggestion on umask. Will include in the next rev.

Regarding the home directory, the script assumes that the user home is created because PAM executes before the system-auth triggering home creation.

Reply:
> If the home directory is not created for some unexpected failures, this script is not expected to take this responsibility so it is expected to fail and it would fail when trying to create the .ssh directory.

I think that it a good point since, as you explained, the home creation should be a required PAM module

> Are you suggesting to introduce an extra verification step that verifies that HOME is created and fails otherwise?
I thin k it would be good and would fail with a more clear error message

yes I think that could be useful.

Reply:
Added